### ISSUE TITLE
sql: Schema change transactions deadlock on retryable error in 1.1.6

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#23979]
A transaction that meets the following criteria will experience a kind of deadlock in CockroachDB 1.1.6:
- Uses the `SAVEPOINT cockroach_restart` protocol
- Contains a schema change that requires a backfill (such as index creation, but not table creation)
- That schema change is not the first statement in the transaction (which prevents automatic server-side retries
- A retryable error is generated for the `RELEASE SAVEPOINT` statement. When the deadlock occurs, the client will see `RELEASE SAVEPOINT` hang without returning anything.

This deadlock leaves the transaction's heartbeat loop running, so it cannot be aborted by other normal-priority transactions. This leaves a pending intent in the system tables which will cause cluster-wide problems. The situation can be resolved by killing the gateway node for the deadlocked transaction, or (I think) running a `PRIORITY HIGH` `SELECT` over the `system.descriptor` table. 

The root cause is an error in the SQL transaction state machine logic for the state that follows the `RELEASE SAVEPOINT` statement. A successful `RELEASE SAVEPOINT` ends the transaction; an unsuccessful one leaves it open. The schema change logic was incorrectly triggering after an unsuccessful `RELEASE SAVEPOINT`, and proceeded to immediately deadlock against the still-open transaction. (our deadlock detection didn't detect this because it's not a deadlock in SQL transactions; one part of the deadlock is the server goroutine that we're blocking). I think the fix should be a [one-liner](https://github.com/cockroachdb/cockroach/blob/8b05735a96d389ecda54a4f46d83c1fbfbd55ea2/pkg/sql/executor.go#L902)

This transaction state machine code has been completely rewritten (by @andreimatei) for 2.0. I haven't yet determined whether the bug is still present there. (I don't think my current test will work as-is because retryable errors have gotten a lot less common too)

**Comment by bdarnell** [comment]
`master` is unaffected by default, but the same bug is present with `COCKROACH_USE_LEGACY_CONN_HANDLING=true`

**Comment by andreimatei** [comment]
`master` is indeed good - we only run schema changes when we've committed the KV txn:
https://github.com/cockroachdb/cockroach/blob/037fd80b2e033de4415bc5beffce5a250524dc60/pkg/sql/conn_executor.go#L1657

Don't worry about `COCKROACH_USE_LEGACY_CONN_HANDLING`. It's already pretty broken and I'm currently deleting it along with all the old code.

Excellent find, Ben! How exactly did you find it?

**Comment by andreimatei** [comment]
Another thing that 2.0 has that would have helped here is that closing of client connections is handled better - if the client goes away, it's more likely that that would have stopped the runaway heartbeat.

**Comment by bdarnell** [comment]
> How exactly did you find it?

Lots of staring at stack traces and applying the [Feynman algorithm](http://wiki.c2.com/?FeynmanAlgorithm). The obvious problem is hundreds of goroutines waiting to push a transaction. Initially I was looking for bugs in the `pushTxnQueue`, but that went nowhere. `panicparse` on the stack traces showed that there were only two distinct things going on at the SQL level: Dozens of queries to virtual schemas (this was `SHOW TABLES` commands), and a single transaction blocked in `execSchemaChange` and `AcquireLease`. 

From there, I started with the hypothesis that the pushTxnQueue is working properly, and some transaction is being held open, blocking everything else. There is a single open `TxnCoordSender.heartbeatLoop` in the cluster. The only writing transaction that we have any evidence for is that schema change, so either that one is still open or a heartbeat loop managed to survive its transaction disappearing without any other trace. Then I realized that from the logs the user supplied, all of this was using `SAVEPOINT`, and that seemed like something under-tested in the executor so I started tracing through the surrounding code.
